server-spi-private/src/main/java/org/keycloak/authorization/policy/evaluation/PolicyQueryBuilder.java

```diff
@@ -0,0 +1,129 @@
+        return new PolicyQueryBuilder(resourceServer, policyStore, permission, true);
+    }
+
+    /**
+     * This method kicks off the evaluation that is configured at the end of the chain.
+     * Because the provided policy consumers will also house the results in some way
+     * (for instance, a decision cache, set, collection, or other output...)
+     * we are only concerned with figuring out what we have configured and then executing
+     *
+     * @return if and only if no policy consumer was provided for that query returns a Set of resultant policies mapped by the query name
+     */
+    public Map<String, List<Policy>> query() {
+        Map<String, List<Policy>> results = new HashMap<>();
+
+        if(Boolean.TRUE.equals(isFullSearch)) {
+            // get scope policies that are explicitly not linked to any resources
+            if(scopes != null) {
+                if(scopeConsumer != null) {
+                    policyStore.findByScopes(resourceServer, null, new LinkedList<>(scopes), scopeConsumer);
+                } else {
+                    results.put("scope", policyStore.findByScopes(resourceServer, null, new LinkedList<>(scopes)));
+                }
+            }
+
+            if(resource != null) {
+                // get type policies that are explicitly not linked to any resources
+                if(resource.getType() != null) {
+                    if(typeConsumer != null) {
+                        policyStore.findByResourceType(resourceServer, false, resource.getType(), typeConsumer);
+                    } else {
+                        ArrayList<Policy> typePolicies = new ArrayList<>();
+                        policyStore.findByResourceType(resourceServer, false, resource.getType(), typePolicies::add);
+                        results.put("type", typePolicies);
+                    }
+                }
+
+                // get resource policies that are explicitly not linked to any scopes but are to the resource
+                if(resourceConsumer != null) {
+                    policyStore.findByResource(resourceServer, false, resource, resourceConsumer);
+                } else {
+                    ArrayList<Policy> resourcePolicies = new ArrayList<>();
+                    policyStore.findByResource(resourceServer, false, resource, resourcePolicies::add);
+                    results.put("resource", resourcePolicies);
+                }
+            }
+        }
+
+        // find only policies that are explicitly linked to the provided resource AND scopes
+        if(scopes != null && resource != null) {
+            if(resourceScopeConsumer != null) {
+                policyStore.findByScopes(resourceServer, resource, new LinkedList<>(scopes), resourceScopeConsumer);
+            } else {
+                results.put("resource_scope", policyStore.findByScopes(resourceServer, resource, new LinkedList<>(scopes)));
+            }
+        }
+
+        return results;
```

Comment:
I might lack context here, but the default would be an empty results here. I wonder if there are combinations where one would just throw an exception because not all required elements were filled / the build was incomplete. This would help developers to find the error in their ways earlier.

Reply:
This section of the code behaves a bit different than typical java flows and more like a functional programming flow. The policy store methods take a consumer that executes after the results are found and that's what determines what happens to the resultant records from the DB. I tried to set this up in a way where you start with the empty results map and records only get added to it if you don't configure some other consumer. 

It's a bit confusing to me if I'm being honest, but I suspect that it's partially because this section of the code works a bit more like javascript callback methods and middleware

Reply:
Thinking about it, I don't think I explicitly answered the question. The intended behavior is that if you were to use this class by just constructing it and then calling "query" the default behavior would be to return the mapped results with no extra configuration needed. That effectively gives a wrapper around the methods that makes it work at least somewhat similarly to how the other storage methods work in other sections of the code. 

I do agree that it could be cleaner to just throw an exception though. It's not that much extra effort to expect the user to provide their own consumer that can do the result mapping

Reply:
Should fine now as we don't the policy builder class anymore.